File: July/Day20/delete-duplicate-folders-in-system.java
class Trie {
    String serial = "";
    Map<String, Trie> children = new HashMap<>();
}

class Solution {
    Map<String, Integer> freq = new HashMap<>();
    List<List<String>> result = new ArrayList<>();

    public List<List<String>> deleteDuplicateFolder(List<List<String>> paths) {
        Trie root = new Trie();

        for (List<String> path : paths) {
            Trie curr = root;
            for (String folder : path) {
                curr.children.putIfAbsent(folder, new Trie());
                curr = curr.children.get(folder);
            }
        }

        construct(root);

        dfs(root, new ArrayList<>());
        return result;
    }

    private void construct(Trie node) {
        if (node.children.isEmpty()) return;

        List<String> v = new ArrayList<>();
        for (Map.Entry<String, Trie> entry : node.children.entrySet()) {
            construct(entry.getValue());
            v.add(entry.getKey() + "(" + entry.getValue().serial + ")");
        }
        Collections.sort(v);
        node.serial = String.join("", v);
        freq.put(node.serial, freq.getOrDefault(node.serial, 0) + 1);
    }

    private void dfs(Trie node, List<String> path) {
        if (freq.getOrDefault(node.serial, 0) > 1) return;
        if (!path.isEmpty()) result.add(new ArrayList<>(path));

        for (Map.Entry<String, Trie> entry : node.children.entrySet()) {
            path.add(entry.getKey());
            dfs(entry.getValue(), path);
            path.remove(path.size() - 1);
        }
    }
}

File: July/Day20/delete-duplicate-folders-in-system.cc
class Trie {
public:
    string serial;
    unordered_map<string, Trie*> children;

    Trie() : serial("") {}
};

class Solution {
public:
    vector<vector<string>> deleteDuplicateFolder(vector<vector<string>>& paths) {
        Trie* root = new Trie();

        for (auto& path : paths) {
            Trie* curr = root;
            for (auto& folder : path) {
                if (!curr->children.count(folder)) {
                    curr->children[folder] = new Trie();
                }
                curr = curr->children[folder];
            }
        }

        unordered_map<string, int> freq;

        function<void(Trie*)> construct = [&](Trie* node) {
            if (node->children.empty()) return;

            vector<string> v;
            for (auto& [folder, child] : node->children) {
                construct(child);
                v.push_back(folder + "(" + child->serial + ")");
            }
            sort(v.begin(), v.end());
            node->serial = accumulate(v.begin(), v.end(), string(""));
            freq[node->serial]++;
        };

        construct(root);

        vector<vector<string>> result;
        vector<string> path;

        function<void(Trie*)> collect = [&](Trie* node) {
            if (freq[node->serial] > 1) return;
            if (!path.empty()) result.push_back(path);

            for (auto& [folder, child] : node->children) {
                path.push_back(folder);
                collect(child);
                path.pop_back();
            }
        };

        collect(root);
        return result;
    }
};